### PR TITLE
Rust: Add support for the errors

### LINKED ISSUE #175: Support Everest Errors Raising and handling in Rust
Everest Framework supports errors that can be raised and handled by modules.

Rust Implementation is missing this feature. 
Side note: We may need to wait until the concept of errors will stabilize, since it is still work in progress.

### PR DESCRIPTION
**Pull request by dorezyuk**
Adds error support for rust. 
Closes #175

### CODE REVIEW

File: everestrs/everestrs-build/jinja/errors.jinja2 [hunk 1]
```diff
@@ -0,0 +1,27 @@
+{%- for name, errors in involved_errors | items  %}
+    pub mod {{ name | snake }} {
```
Comment (ju4nlu) [inline everestrs/everestrs-build/jinja/errors.jinja2:2]: Seems like all generated lines from this file come with an extra indentation, isn't it the case?

File: everestrs/everestrs-build/jinja/module.jinja2 [hunk 2]
```diff
@@ -211,6 +216,32 @@ impl ::everestrs::Subscriber for Module {
...
+{% for req in requires %}
+            "{{ req.implementation_id }}" => {
+    {% if req.min_connections == 1 and req.max_connections == 1 %}
+                dispatch_error_to_{{ req.interface | snake }}(&context, self.{{ req.implementation_id  | identifier }}.as_ref(), error, raised)
+    {% else %}
+                dispatch_error_to_{{ req.interface | snake }}(&context, self.{{ req.implementation_id  | identifier }}_slots[index].as_ref(), error, raised)
+    {% endif %}
+            },
+{% endfor %}
+            _ => eprintln!("Received an unknonw error from {implementation_id}"),
```
Comment (ju4nlu) [inline everestrs/everestrs-build/jinja/module.jinja2]: typo
```suggestion
            _ => eprintln!("Received an unknown error from {implementation_id}"),
```

File: everestrs/everestrs-build/jinja/module.jinja2 [hunk 3]
```diff
@@ -211,6 +216,32 @@ impl ::everestrs::Subscriber for Module {
...
+        error: ::everestrs::ErrorType,
+        raised: bool
+    ) {
+        let context = Context {
+            publisher: &self.publisher,
+            name: implementation_id,
+            index,
+        };
+        match implementation_id {
+{% for req in requires %}
```
Comment (ju4nlu) [inline everestrs/everestrs-build/jinja/module.jinja2:232]: indenting the jinja lines at the same level as the rust code will improve the reading experience

File: everestrs/everestrs-build/src/codegen.rs [hunk 4]
```diff
@@ -307,12 +315,145 @@ impl CommandContext {
...
+            let path = splits.next().ok_or(anyhow::anyhow!("No path defined"))?;
+
+            // Split the path and remove the empty parts.
+            // (The first element might be empty if we have a leading `/`).
+            let paths = path
+                .split("/")
+                .filter(|path| !path.is_empty())
+                .collect::<Vec<_>>();
+
+            anyhow::ensure!(paths.len() == 2);
```
Comment (ju4nlu) [inline everestrs/everestrs-build/src/codegen.rs]: Is the message printed if this is not fulfilled meaningful enough? I believe it will just print `Error: Condition failed: `paths.len() == 2` (1 vs 2)`. Same happens with the line below this one

File: everestrs/everestrs-build/src/codegen.rs [hunk 5]
```diff
@@ -651,3 +811,46 @@ pub fn emit(manifest_path: PathBuf, everest_core: Vec<PathBuf>) -> Result<String
...
 }
+
+#[cfg(test)]
+mod tests {
+    #[test]
+    fn test_split_paths_invalid() {
+        use super::impl_error::*;
+        let invalid_input = [
+            "/foo/bar/baz", // too many
+            "/foo",         // too view,
```
Comment (ju4nlu) [inline everestrs/everestrs-build/src/codegen.rs]:
```suggestion
            "/foo",         // too few,
```

I guess

File: schemas/error-declaration-list.yaml [hunk 6]
```diff
@@ -21,10 +21,6 @@ $defs:
       description:
         type: string
         minLength: 2
-      namespace:
-        type: string
-        minLength: 2
-        pattern: ^[a-z][a-z0-9]*$
```
Comment (ju4nlu) [inline schemas/error-declaration-list.yaml:27]: Not really related to supporting errors in EVerest, I think this should be a part of a different MR

Comment (dorezyuk) [reply schemas/error-declaration-list.yaml:27]: I've discussed this with @andistorm - this just updates the schema

File: everestrs/everestrs-build/jinja/errors.jinja2 (hunk 1 above)
Comment (dorezyuk) [reply everestrs/everestrs-build/jinja/errors.jinja2:2]: these lines are include in the errors module in the the module.jinja file - so they must be indented

File: everestrs/everestrs-build/jinja/module.jinja2 (hunk 3 above)
Comment (dorezyuk) [reply everestrs/everestrs-build/jinja/module.jinja2:232]: we did it like this mostly through out the file indenting rust and jinda independently

**Review by SirVer — APPROVED** [review]
